opt_einsum/paths.py

```diff
@@ -88,6 +89,62 @@ def optimal(input_sets, output_set, idx_dict, memory_limit):
     return path
 
 
+def _calc_k12_flops(inputs, output, remaining, i, j, size_dict):
+    k1, k2 = inputs[i], inputs[j]
+    either = k1 | k2
+    two = k1 & k2
+    one = either - two
+    keep = frozenset.union(output, *(inputs[p] for p in remaining if p not in (i, j)))
+
+    k12 = one | (two & keep)
+    cost = helpers.flop_count(either, two - keep, 2, size_dict)
+    return k12, cost
+
+
+def roptimal(inputs, output, size_dict, memory_limit=math.inf):
+
+    inputs = tuple(map(frozenset, inputs))
+    output = frozenset(output)
+
+    best = {'flops': math.inf, 'path': (tuple(range(len(inputs))),)}
+
+    def _iterate(path, remaining, inputs, flops):
+
+        # reached end of path -> assess it
+        if len(remaining) == 1:
+            if flops < best['flops']:
+                best['flops'] = flops
+                best['path'] = path
+            return
+
+        # check all possible remaining paths
+        for i, j in itertools.combinations(remaining, 2):
+            k12, flops12 = _calc_k12_flops(inputs, output, remaining, i, j, size_dict)
+
+            # sieve based on current best flops
+            new_flops = flops + flops12
+            if new_flops > best['flops']:
+                continue
+
+            # sieve based on memory limit
+            if memory_limit is not math.inf:
```

Comment:
note `math.inf` is not defined in Python 2. Also you shouldn't use `is` to compare floats. A safer version is:
```py
if memory_limit < float('inf'):
```

Reply:
Ah thanks, my python2 knowledge is quite lacking...